src/main/java/org/testng/internal/Parameters.java

```diff
@@ -427,18 +428,51 @@ public static ParameterHolder handleParameters(ITestNGMethod testMethod,
         allParameterNames.put(n, n);
       }
 
-      parameters = MethodInvocationHelper.invokeDataProvider(
+      final Iterator<Object[]> parameters = MethodInvocationHelper.invokeDataProvider(
           dataProviderHolder.instance, /* a test instance or null if the dataprovider is static*/
           dataProviderHolder.method,
           testMethod,
           methodParams.context,
           fedInstance,
           annotationFinder);
 
-      Iterator<Object[]> filteredParameters = filterParameters(parameters,
```

Comment:
This is the most important part of the PR: 

We are not allowed to iterate on the iterator here because:
- When users are using Iterator, they expect a lazy behavior, and
- Iterator data providers may fail with exception that are not well treated here.